File: benchmarks/cross-language/c/skynet.c
/**
 * C Skynet Benchmark
 * Based on https://github.com/atemerev/skynet
 * Uses pthreads while the subtree is larger than SEQ_THRESHOLD, sequential below.
 * Spawning 1M OS threads is not feasible; limits concurrent threads to ~1000.
 *
 * Compile with: gcc -O3 -march=native skynet.c -o skynet -pthread
 */

#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>
#include <string.h>
#include <time.h>

/* Sequential below SEQ_THRESHOLD. Same threshold in every language in this
 * suite, so all of them create the same 1,111 concurrency units and perform the
 * same 1,000,000 leaf additions. */
#define SEQ_THRESHOLD 1000

static long long skynet_seq(long long offset, long long size) {
    if (size == 1) return offset;
    long long child_size = size / 10;
    long long sum = 0;
    for (int i = 0; i < 10; i++) {
        sum += skynet_seq(offset + (long long)i * child_size, child_size);
    }
    return sum;
}

typedef struct {
    long long offset;
    long long size;
    int depth;
    long long result;
} SkynetArg;

static void* skynet_thread(void* arg) {
    SkynetArg* a = (SkynetArg*)arg;
    long long offset = a->offset;
    long long size   = a->size;
    int       depth  = a->depth;

    if (size <= SEQ_THRESHOLD) {
        a->result = skynet_seq(offset, size);
        return NULL;
    }

    long long child_size = size / 10;
    SkynetArg children[10];
    pthread_t threads[10];

    for (int i = 0; i < 10; i++) {
        children[i].offset = offset + (long long)i * child_size;
        children[i].size   = child_size;
        children[i].depth  = depth + 1;
        children[i].result = 0;
        pthread_create(&threads[i], NULL, skynet_thread, &children[i]);
    }

    long long sum = 0;
    for (int i = 0; i < 10; i++) {
        pthread_join(threads[i], NULL);
        sum += children[i].result;
    }
    a->result = sum;
    return NULL;
}

static long long get_leaves(void) {
    const char* env = getenv("SKYNET_LEAVES");
    if (env) return atoll(env);
    env = getenv("BENCHMARK_MESSAGES");
    if (env) return atoll(env);
    return 1000000LL;
}

int main(void) {
    long long num_leaves = get_leaves();

    /* Divide by the concurrency units created, not by the tree's node count.
     *
     * Every language in this suite now uses the same SEQ_THRESHOLD, so every
     * one creates the same 1,111 units and performs the same num_leaves leaf
     * additions. A per-unit cost is therefore directly comparable, and it is
     * what skynet is for: the price of creating a unit, passing its result up
     * and aggregating.
     *
     * It used to divide by the full 1,111,111-node tree while creating between
     * 1,111 and 1,111,111 units depending on the language, so whichever
     * implementation created the fewest scored highest. */
    long long total_actors = 1;
    for (long long n = num_leaves; n > SEQ_THRESHOLD; n /= 10) {
        total_actors += n / SEQ_THRESHOLD;
    }
    const long long rate_base = total_actors;

    printf("=== C Skynet Benchmark ===\n");
    printf("Leaves: %lld, concurrency units: %lld (sequential below %d)\n\n",
           num_leaves, total_actors, SEQ_THRESHOLD);

    SkynetArg root = { .offset = 0, .size = num_leaves, .depth = 0, .result = 0 };

    struct timespec ts_start, ts_end;
    clock_gettime(CLOCK_MONOTONIC, &ts_start);

    pthread_t root_thread;
    pthread_create(&root_thread, NULL, skynet_thread, &root);
    pthread_join(root_thread, NULL);

    clock_gettime(CLOCK_MONOTONIC, &ts_end);

    long long elapsed_ns = (long long)(ts_end.tv_sec - ts_start.tv_sec) * 1000000000LL
                         + (ts_end.tv_nsec - ts_start.tv_nsec);
    long long elapsed_us = elapsed_ns / 1000;

    printf("Sum: %lld\n", root.result);
    if (elapsed_us > 0) {
        long long ns_per_msg   = elapsed_ns / rate_base;
        long long throughput_m = rate_base / elapsed_us;
        long long leftover     = rate_base - (throughput_m * elapsed_us);
        long long frac         = (leftover * 100) / elapsed_us;
        printf("ns/msg:         %lld\n", ns_per_msg);
        printf("Throughput:     %lld.%02lld M msg/sec\n", throughput_m, frac);
    }
    return 0;
}

File: benchmarks/cross-language/zig/skynet.zig
// Zig Skynet Benchmark
// Based on https://github.com/atemerev/skynet
// Uses std.Thread while the subtree is larger than SEQ_THRESHOLD.
// Spawning 1M OS threads is not feasible; limits concurrent threads to ~1000.

const std = @import("std");
const Thread = std.Thread;
const print = std.debug.print;

const SEQ_THRESHOLD: i64 = 1000;

fn getLeaves() i64 {
    if (std.posix.getenv("SKYNET_LEAVES")) |val| {
        return std.fmt.parseInt(i64, val, 10) catch 1_000_000;
    }
    if (std.posix.getenv("BENCHMARK_MESSAGES")) |val| {
        return std.fmt.parseInt(i64, val, 10) catch 1_000_000;
    }
    return 1_000_000;
}

fn skynetSeq(offset: i64, size: i64) i64 {
    if (size == 1) return offset;
    const child_size = @divTrunc(size, 10);
    var sum: i64 = 0;
    var i: i64 = 0;
    while (i < 10) : (i += 1) {
        sum += skynetSeq(offset + i * child_size, child_size);
    }
    return sum;
}

const SkynetArg = struct {
    offset: i64,
    size: i64,
    depth: usize,
    result: i64 = 0,
};

fn skynetThread(arg: *SkynetArg) void {
    const offset = arg.offset;
    const size = arg.size;
    const depth = arg.depth;

    if (size <= SEQ_THRESHOLD) {
        arg.result = skynetSeq(offset, size);
        return;
    }

    const child_size = @divTrunc(size, 10);
    var children: [10]SkynetArg = undefined;
    var threads: [10]Thread = undefined;

    for (0..10) |i| {
        children[i] = SkynetArg{
            .offset = offset + @as(i64, @intCast(i)) * child_size,
            .size = child_size,
            .depth = depth + 1,
        };
        threads[i] = Thread.spawn(.{}, skynetThread, .{&children[i]}) catch {
            // Fallback: compute sequentially if spawn fails
            children[i].result = skynetSeq(children[i].offset, children[i].size);
            threads[i] = undefined;
            continue;
        };
    }

    var sum: i64 = 0;
    for (0..10) |i| {
        threads[i].join();
        sum += children[i].result;
    }
    arg.result = sum;
}

fn getTimeNs() u64 {
    const ts = std.posix.clock_gettime(std.posix.CLOCK.MONOTONIC) catch return 0;
    return @as(u64, @intCast(ts.sec)) * 1_000_000_000 + @as(u64, @intCast(ts.nsec));
}

pub fn main() !void {
    const num_leaves = getLeaves();

    // Concurrency units actually created; also the divisor, since every language
    // in this suite creates the same count.
    var total_actors: i64 = 1;
    var n = num_leaves;
    while (n > SEQ_THRESHOLD) : (n = @divTrunc(n, 10)) {
        total_actors += @divTrunc(n, SEQ_THRESHOLD);
    }

    print("=== Zig Skynet Benchmark ===\n", .{});
    print("Leaves: {}, concurrency units: {} (sequential below {})\n\n", .{ num_leaves, total_actors, SEQ_THRESHOLD });

    var root = SkynetArg{ .offset = 0, .size = num_leaves, .depth = 0 };

    const start = getTimeNs();
    const root_thread = try Thread.spawn(.{}, skynetThread, .{&root});
    root_thread.join();
    const end = getTimeNs();

    const elapsed_ns = @as(i64, @intCast(end - start));
    const elapsed_us = @divTrunc(elapsed_ns, 1000);

    print("Sum: {}\n", .{root.result});
    if (elapsed_us > 0) {
        const ns_per_msg = @divTrunc(elapsed_ns, total_actors);
        // Float, like the other four Zig benchmarks. The manual
        // integer-and-fraction version printed "0.+4" for any rate below
        // 1 M/sec, which the runner's parser reads as garbage.
        const throughput = @as(f64, @floatFromInt(total_actors)) /
            (@as(f64, @floatFromInt(elapsed_ns)) / 1_000_000_000.0) / 1_000_000.0;
        print("ns/msg:         {}\n", .{ns_per_msg});
        print("Throughput:     {d:.2} M msg/sec\n", .{throughput});
    }
}
